SMU.cpp

```diff
@@ -90,30 +92,57 @@ void SessionItem::start()
       for (auto sig: chan->m_signals) {
         sig->m_buffer->setRate(1.0/m_sample_rate);
         sig->m_buffer->allocate(m_sample_count);
-        sig->m_signal->measure_buffer(sig->m_buffer->data(), m_sample_count);
+
+        if (m_continuous) {
+          sig->m_signal->measure_callback([=](float d){
+            sig->m_buffer->shift(d);
+          });
+
+          connect(sig->m_src, &SrcItem::changed, [=] {
+            dev->m_device->lock();
+            sig->m_src->update();
+            dev->m_device->unlock();
+          });
+        } else {
+          sig->m_signal->measure_buffer(sig->m_buffer->data(), m_sample_count);
+        }
         sig->m_src->update();
+
       }
     }
   }
 
-  m_session->start(m_sample_count);
+  m_session->start(continuous ? 0 : m_sample_count);
 }
 
-void SessionItem::onAttached()
+void SessionItem::onAttached(Device *device)
 {
   qDebug() << "attached\n";
+  auto dev = m_session->add_device(device);
+  m_devices.append(new DeviceItem(this, device));
+  devicesChanged();
+}
+
+void SessionItem::onDetached(Device* device){
+  qDebug() << "detached\n";
   qDebug() << m_devices;
-  for (auto i: m_session->m_available_devices) {
-    auto dev = m_session->add_device(&*i);
-    m_devices.append(new DeviceItem(this, dev));
+  if (!m_active) {
+      m_session->remove_device(device);
+  }
+  else {
+      this->cancel();
```

Comment:
need to `wait` for the transfers to actually cancel. Also, don't you still need to then call remove_device?